### DISCUSSION TITLE
hono validator adding File type to every form field

### DISCUSSION THREAD

**Discussion by jwhits** [honojs/hono#2604] [Q&A]
When using the following code from the stacks example on the hono website

```javascript
// functions/api/[[route]].ts
import { Hono } from 'hono'
import { handle } from 'hono/cloudflare-pages'
import { z } from 'zod'
import { zValidator } from '@hono/zod-validator'

const app = new Hono()

const schema = z.object({
  id: z.string(),
  title: z.string(),
})

type Todo = z.infer<typeof schema>

const todos: Todo[] = []

const route = app
  .post('/todo', zValidator('form', schema), (c) => {
    const todo = c.req.valid('form')
    todos.push(todo)
    return c.json({
      message: 'created!',
    })
  })
  .get((c) => {
    return c.json({
      todos,
    })
  })

export type AppType = typeof route
```
zValidator (hono validator) is adding '| File' type to each one of the form fields. So when using RPC on the frontend you end up with the following types.
```javascript
const $post: (args: {
    form: {
        id: string | File;
        title: string | File;
    };
...
```
Is there a reason you would want every form field to be a string or a File? I would have thought you would specify exactly what you want?

**Comment by clyxudev** [comment]
Yes, that typing comes from how `form` validation is modeled in Hono.

A `form` request is parsed through the platform `FormData` API, and `FormData` values can be text fields or uploaded files. In Hono's types this is represented as a form record whose values are `string | File` on the parsed side. So the RPC client has to be conservative: a form field might be a file even if your Zod schema later narrows the validated output to strings.

At runtime your schema still validates the fields as strings:

```ts
const schema = z.object({
  id: z.string(),
  title: z.string(),
})
```

and inside the handler:

```ts
const todo = c.req.valid('form')
```

`todo` should be the validated Zod output.

If the frontend payload is purely structured data and not an actual HTML form/file upload, the cleaner path is usually to use `json` instead of `form`:

```ts
.post('/todo', zValidator('json', schema), (c) => {
  const todo = c.req.valid('json')
  return c.json({ message: 'created!' })
})
```

That gives the RPC input the shape you probably expect for text-only data, without the `File` union. Use `form` when you want browser `FormData` semantics; use `json` when you want a typed object payload.
